gnrc_tcp: check if option length field is present before accessing it

nmeum opened the pull request:
TCP options have a maximum size of 3 byte (kind, length, value). The
current code only checks for the presence of the first field. Before
accessing the second field (length) the code must ensure that a length
field is even present.

**Discussion:** Change is somewhat ugly as it introduces a magic constant.
The proper way of solving this would probably be not using a packed struct
to parse these variable-length options.

brummer-simon commented:
Hi nmeum,

Good catch again. First of all TCP options do not have a length maximum size of three bytes as stated in the PR descriptions.

Supported Options introduced with the original TCP have either a single Byte (EOL and NOP) or four Bytes (MSS). Additional TCP extensions introduce options with variable length. That been said you are right, the length field must be checked before access. 

Now to your PR. The Logic is fine, but usage of a magic numbers 
goes against the RIOT Styleguide. However you could place a constant in 'sys/include/net/tcp.h' and use it in the affected code segment.

If you change that you get my ACK.

Cheers SImon

nmeum commented:
> Good catch again. First of all TCP options do not have a length maximum size of three bytes as stated in the PR descriptions.

Thanks for pointing that out. I wanted to refer to fields, not bytes. Changed the commit message accordingly.

>  Now to your PR. The Logic is fine, but usage of a magic numbers
goes against the RIOT Styleguide. However you could place a constant in 'sys/include/net/tcp.h' and use it in the affected code segment.

Thanks for the quick feedback. Added a constant in the file you referenced. I still think that's a bad idea to parse variable-length options with a packed struct though, but maybe that can be addressed separately.

brummer-simon commented:
> Thanks for the quick feedback. Added a constant in the file you referenced. I still think that's a bad idea to parse variable-length options with a packed struct though, but maybe that can be addressed separately.

After test rewriting is merged and gnrc_tcp is integrated into the sock interface, I though about refactoring some parts of gnrc_tcp. Maybe rewriting the option parsing could be part of it.
